core/src/main/java/io/questdb/griffin/SqlCompiler.java

```diff
@@ -1796,84 +1797,92 @@ private CompiledQuery insert(ExecutionModel executionModel, SqlExecutionContext
-                                model,
-                                valueFunctions,
-                                metadata,
-                                writerTimestampIndex,
-                                i,
-                                index,
-                                function,
-                                node.position,
-                                executionContext.getBindVariableService()
-                        );
+                if (columnSetSize > 0) {
+                    valueFunctions = new ObjList<>(columnSetSize);
+                    for (int i = 0; i < columnSetSize; i++) {
+                        int index = metadata.getColumnIndexQuiet(columnSet.get(i));
+                        if (index > -1) {
+                            final ExpressionNode node = model.getRowValues(t).getQuick(i);
+
+                                Function function = functionParser.parseFunction(
+                                        node,
+                                        GenericRecordMetadata.EMPTY,
+                                        executionContext
+                                );
+
+                                function = validateAndConsume(
+                                        model,
+                                        t,
+                                        valueFunctions,
+                                        metadata,
+                                        writerTimestampIndex,
+                                        i,
+                                        index,
+                                        function,
+                                        node.position,
+                                        executionContext.getBindVariableService()
+                                );
+
+                                if (writerTimestampIndex == index) {
+                                    timestampFunction = function;
+                                }
 
-                        if (writerTimestampIndex == index) {
-                            timestampFunction = function;
+                            } else {
+                                throw SqlException.invalidColumn(model.getColumnPosition(i), columnSet.get(i));
+                            }
                         }
-
                     } else {
-                        throw SqlException.invalidColumn(model.getColumnPosition(i), columnSet.get(i));
+                        final int columnCount = metadata.getColumnCount();
+                        final ObjList<ExpressionNode> values = model.getRowValues(t);
+                        final int valueCount = values.size();
+                        if (columnCount != valueCount) {
+                            throw SqlException.$(
+                                    model.getEndOfValuesPosition(),
+                                    "not enough values [expected=").put(columnCount).put(", actual=").put(values.size())
```

Comment:
this error message is factually incorrect when row group has more values than target columns